Correct error in process DFE experiment generation.

kylegulshen opened the pull request:
Description
-----------

The old code was not properly generating all of the required experiment settings for process DFE. In particular only 7^n of the required (4^n - 1)2^n settings were included. For a single qubit all settings were present. For two qubits there was one missing setting for each observable with exactly one Identity term. 

The error is illustrated by the following mismatch between the analytically calculated fidelity and the simulated fidelity between the identity and CPHASE(theta) for various theta:
![image](https://user-images.githubusercontent.com/41916921/69570065-03fdee00-0f8d-11ea-9537-0b106ca535eb.png)


Checklist
---------

- [x] The above description motivates these changes.
- [x] There is a unit test that covers these changes.
- [x] The code respects the API separation discussed in .github/CONTRIBUTING.md.
- [x] Relevant references and equations are cited using our standard reference style.
- [ ] All new and existing tests pass locally and on Semaphore.
- [x] Parameters have type hints with [PEP 484 syntax](https://www.python.org/dev/peps/pep-0484/).
- [x] Functions and classes have useful sphinx-style docstrings.
- [ ] (New Feature) The docs and example notebooks have been updated accordingly.
- [ ] (Bugfix) The associated issue is referenced above using
      [auto-close keywords](https://help.github.com/en/articles/closing-issues-using-keywords).
- [ ] The changelog (`docs/source/changes.rst`) has a description of this change.

caryan commented on forest/benchmarking/direct_fidelity_estimation.py:
```diff
@@ -112,10 +113,24 @@ def generate_exhaustive_process_dfe_experiment(benchmarker: BenchmarkConnection,
+            in_state = functools.reduce(mul,
+                                        (TensorProductState((_OneQState(state_labels[idx], sign,
+                                                                        qubit),))
+                                         for idx, (qubit, sign) in enumerate(zip(qubits,
```
you can zip over `state_labels` as well

caryan commented on forest/benchmarking/direct_fidelity_estimation.py:
```diff
@@ -112,10 +113,24 @@ def generate_exhaustive_process_dfe_experiment(benchmarker: BenchmarkConnection,
+        non_identity_idx = [0 if label == 'I' else 1 for label in pauli_labels]
+        state_labels = ['Z' if label == 'I' else label for label in pauli_labels]
+        for eigenstate in itertools.product([0, 1], repeat=len(qubits)):
+            in_state = functools.reduce(mul,
```
this assignment can be simplified by directly passing the generator to `TensorProductState`'s initializer

```suggestion
            in_state = TensorProductState(_OneQState(l, s, q) for l,s,q in zip(state_labels, eigenstate, qubits))
```

kylegulshen commented:
Thanks for the comments!  

I addressed those comments and then realized a couple of other things should have been updated as well. 

1) state dfe can be simplified significantly after the changes
2) monte carlo needed to be updated as well
3) the changes made some of the content of the example notebook wrong
4) I noticed a couple other minor mistakes in the example notebook explanation and bad formatting in the code

If it is preferable that I roll back the changes to address your comments only and open new PRs to address the other stuff, let me know.

kylegulshen commented:
I believe #205 should fix the test issue, but it can't be merged until a separate issue is resolved.
